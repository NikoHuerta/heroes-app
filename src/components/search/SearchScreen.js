import React, { useMemo } from 'react';
import queryString from 'query-string';
import { useNavigate, useLocation } from 'react-router-dom';
import { useForm } from '../../hooks/useForm';
import { HeroCard } from '../heroes/HeroCard';
import { getHeroesByName } from '../../selectors/getHeroByName';

export const SearchScreen = () => {

    const location = useLocation();
    const { q = '' } = queryString.parse(location.search);

    const navigate = useNavigate();
    const [{ searchString }, handleInputChange] = useForm({ searchString: q });

    const heroesFiltered = useMemo( () => getHeroesByName( q ), [q] );

    const handleSearch = (e) =>{
        e.preventDefault();
        if(searchString.trim().length <= 1){
            return ;
        }
        navigate(`?q=${ searchString }`);
    };

    return (
        <div>
            <h1> Search Screen</h1>
            <hr />

            <div className="row">
                <h4>Search Form</h4>
                <hr />
                <form onSubmit={ handleSearch }>
                    <input
                        type="text"
                        placeholder="Find Your Hero"
                        className="form-control"

                        autoComplete="off"
                        name="searchString"
                        onChange={ handleInputChange }
                        value={ searchString }

                        />
                    <button 
                        type="submit"
                        className="btn m-1 btn-block btn-outline-primary"
                    >
                        Search
                    </button>
                </form>
            </div>
            <div>
                <br/>
                <h4>Results</h4>
                <hr/>
                <br/>
                { 
                    (q === '') && <div className="alert alert-info">
                                    Search a Hero
                                </div>
                }

                { 
                    (q !== '' && heroesFiltered.length===0) &&  <div className="alert alert-danger">
                                                                    There is no hero with '{ q }'
                                                                </div>
                }

                <div className="row row-cols-1 row-cols-md-3 g-4 animate__animated animate__fadeIn">
                    {
                        heroesFiltered.map(hero => (
                        <HeroCard key={ hero.id }
                            { ...hero }
                            />
                        ))
                    }
                </div>
            </div>
        </div>
    )
}
