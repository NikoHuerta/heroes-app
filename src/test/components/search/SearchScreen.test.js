import { mount } from "enzyme";
import { SearchScreen } from "../../../components/search/SearchScreen";
import { MemoryRouter } from "react-router-dom";

const mockNavigate = jest.fn();

jest.mock('react-router-dom', () => ({
    ...jest.requireActual('react-router-dom'),
    useNavigate: () => mockNavigate,
}))

describe('Pruebas en <SearchScreen />', () => {

    test('Debe de mostrarse correctamente con valores por defecto', () => {
        
        const wrapper = mount(<MemoryRouter initialEntries={ ['/search'] }>
                                <SearchScreen />
                            </MemoryRouter>);
        expect(wrapper).toMatchSnapshot();
        //expect(wrapper.find('h1').text().trim()).toBe('Search Screen');
        expect(wrapper.find('.alert-info').text().trim()).toBe('Search a Hero');
    });

    test('Debe de mostrar a Batman y el input con el valor del queryString', () => {
        
        const wrapper = mount(<MemoryRouter initialEntries={ ['/search?q=batman'] }>
                                <SearchScreen />
                            </MemoryRouter>);
        
        expect(wrapper.find('input').prop('value')).toBe('batman');
        expect(wrapper).toMatchSnapshot();

    });

    test('Debe de mostrar un error si no se encuentra un heroe', () => {
       
        const searchText = 'batmanError123';
        const wrapper = mount(<MemoryRouter initialEntries={ [`/search?q=${ searchText }`] }>
                                <SearchScreen />
                            </MemoryRouter>);
        
        expect(wrapper.find('.alert-danger').text().trim()).toBe(`There is no hero with '${ searchText }'`);
        expect(wrapper).toMatchSnapshot();

    });


    test('Debe de llamar el navigate al nuevo URL', () => {

        const wrapper = mount(<MemoryRouter initialEntries={ ['/search'] }>
                                <SearchScreen />
                            </MemoryRouter>);
        
        wrapper.find('input').simulate('change', { 
            target: {
                name: 'searchString',
                value: 'batman'
            }
         });  //solo cambia el valor, escribe batman en el input sin disparar el formulario
        
        wrapper.find('form').prop('onSubmit')({
            preventDefault: () => {}
        });

        expect(mockNavigate).toHaveBeenCalled();
        expect(mockNavigate).toHaveBeenCalledWith('?q=batman');



    });
    
    
    

    
    
});
