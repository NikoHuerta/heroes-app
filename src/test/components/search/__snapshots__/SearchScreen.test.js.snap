// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <SearchScreen /> Debe de mostrar a Batman y el input con el valor del queryString 1`] = `
<div>
  <h1>
     Search Screen
  </h1>
  <hr />
  <div
    className="row"
  >
    <h4>
      Search Form
    </h4>
    <hr />
    <form
      onSubmit={[Function]}
    >
      <input
        autoComplete="off"
        className="form-control"
        name="searchString"
        onChange={[Function]}
        placeholder="Find Your Hero"
        type="text"
        value="batman"
      />
      <button
        className="btn m-1 btn-block btn-outline-primary"
        type="submit"
      >
        Search
      </button>
    </form>
  </div>
  <div>
    <br />
    <h4>
      Results
    </h4>
    <hr />
    <br />
    <div
      className="row row-cols-1 row-cols-md-3 g-4"
    >
      <a
        className="my-card"
        href="/hero/dc-batman"
        onClick={[Function]}
      >
        <img
          alt="Batman"
          className="img img-responsive animate__animated animate__fadeIn"
          src="../assets/heroes/dc-batman.jpg"
        />
        <div
          className="profile-name"
        >
          Batman
        </div>
        <div
          className="profile-position"
        >
          Bruce Wayne
        </div>
        <div
          className="profile-overview"
        >
          <div
            className="profile-overview"
          >
            <div
              className="row"
            >
              <div
                className="col-ms-4"
              >
                <h3>
                  DC Comics
                </h3>
                <p>
                  Primera Aparaci√≥n: 
                  <br />
                  Detective Comics #27
                </p>
              </div>
            </div>
          </div>
        </div>
      </a>
    </div>
  </div>
</div>
`;

exports[`Pruebas en <SearchScreen /> Debe de mostrar un error si no se encuentra un heroe 1`] = `
<div>
  <h1>
     Search Screen
  </h1>
  <hr />
  <div
    className="row"
  >
    <h4>
      Search Form
    </h4>
    <hr />
    <form
      onSubmit={[Function]}
    >
      <input
        autoComplete="off"
        className="form-control"
        name="searchString"
        onChange={[Function]}
        placeholder="Find Your Hero"
        type="text"
        value="batmanError123"
      />
      <button
        className="btn m-1 btn-block btn-outline-primary"
        type="submit"
      >
        Search
      </button>
    </form>
  </div>
  <div>
    <br />
    <h4>
      Results
    </h4>
    <hr />
    <br />
    <div
      className="alert alert-danger"
    >
      There is no hero with '
      batmanError123
      '
    </div>
    <div
      className="row row-cols-1 row-cols-md-3 g-4"
    />
  </div>
</div>
`;

exports[`Pruebas en <SearchScreen /> Debe de mostrarse correctamente con valores por defecto 1`] = `
<div>
  <h1>
     Search Screen
  </h1>
  <hr />
  <div
    className="row"
  >
    <h4>
      Search Form
    </h4>
    <hr />
    <form
      onSubmit={[Function]}
    >
      <input
        autoComplete="off"
        className="form-control"
        name="searchString"
        onChange={[Function]}
        placeholder="Find Your Hero"
        type="text"
        value=""
      />
      <button
        className="btn m-1 btn-block btn-outline-primary"
        type="submit"
      >
        Search
      </button>
    </form>
  </div>
  <div>
    <br />
    <h4>
      Results
    </h4>
    <hr />
    <br />
    <div
      className="alert alert-info"
    >
      Search a Hero
    </div>
    <div
      className="row row-cols-1 row-cols-md-3 g-4"
    />
  </div>
</div>
`;
